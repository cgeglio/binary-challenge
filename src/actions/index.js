export const login = user => ({
  type: 'LOGIN',
  user
})

export const logout = user => ({
  type: 'LOGOUT',
  user
})

export const addCards = cards => ({
  type: 'ADD_CARDS',
  cards
})

export const removeCards = cards => ({
  type: 'REMOVE_CARDS',
  cards
})

export const addQuestion = question => ({
  type: 'ADD_QUESTION',
  question
})

export const removeQuestion = question => ({
  type: 'REMOVE_QUESTION',
  question
})

export const addFortune = fortune => ({
  type: 'ADD_FORTUNE',
  fortune
})

export const removeFortune = fortune => ({
  type: 'REMOVE_FORTUNE',
  fortune
})

export const addReading = currentReading => ({
  type: 'ADD_READING',
  currentReading
})

export const removeReading = currentReading => ({
  type: 'REMOVE_READING',
  currentReading
})

export const addFavorite = favorite => ({
  type: 'ADD_FAVORITE',
  favorite
})

export const removeFavorite = favorite => ({
  type: 'REMOVE_FAVORITE',
  favorite
})

export const resetFavorites = favorites => ({
  type: 'RESET_FAVORITES',
  favorites
})
