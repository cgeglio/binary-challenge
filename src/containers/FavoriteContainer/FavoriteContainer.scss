@import "../../mixins.scss";

.favorite-container {
  @include flex();
}
