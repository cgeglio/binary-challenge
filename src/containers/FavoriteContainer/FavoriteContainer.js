import React, { Component } from 'react';
import './FavoriteContainer.scss';
import { connect } from 'react-redux';
import { getCards, getFortune } from '../../apiCalls';
import { addFavorite, removeFavorite, addCards, addFortune } from '../../actions';
import Card from '../../components/Card/Card';
import Loader from '../../components/Loader/Loader';
import save from '../../images/save.png';
import saved from '../../images/saved.png';

class FavoriteContainer extends Component {
  constructor() {
    super();
    this.state={icon: save}
  }

  updateSavedStatus = () => {
    return !this.props.currentReading.saved ? this.saveReading() : this.removeReading();
  }

  saveReading = () => {
    this.setState({icon: saved});
    this.props.currentReading.saved = true;
    this.props.addReadingToFavorites(this.props.currentReading);
  }

  removeReading = () => {
    this.setState({icon: save});
    this.props.currentReading.saved = false;
    this.props.removeReadingFromFavorites(this.props.currentReading);
  }

  determineIcon = () => {
    return this.props.currentReading.saved ? saved : save;
  }

  render() {

    return (
      !this.props.fortune ? <Loader /> :
        <section className='favorite-container'>
          <section className='cards'>
            {this.props.currentReading.cards.map(card => {
              return <Card key={card.name_short} card={card} />
            })}
          </section>
          <section className='reading-details'>
            <button onClick={() => this.updateSavedStatus()} className="save-btn"><img src={this.determineIcon()} alt="save reading icon" className="save-icon"/></button>
            <div>
              <h2 className='question'>{this.props.currentReading.question}</h2>
              <h2>{this.props.currentReading.fortune}</h2>
            </div>
          </section>
      </section>
    )
  }
}

const mapStateToProps = state => ({
  user: state.user,
  cards: state.cards,
  question: state.question,
  fortune: state.fortune,
  currentReading: state.currentReading
})

const mapDispatchToProps = dispatch => ({
  addCardsToStore: cards => (dispatch(addCards(cards))),
  addFortuneToStore: fortune => (dispatch(addFortune(fortune))),
  addReadingToFavorites: favorite => (dispatch(addFavorite(favorite))),
  removeReadingFromFavorites: favorite => (dispatch(removeFavorite(favorite)))
})

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteContainer)
