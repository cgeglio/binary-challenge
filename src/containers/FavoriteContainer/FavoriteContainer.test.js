import React from 'react';
import { FavoriteContainer, mapDispatchToProps, mapStateToProps } from './FavoriteContainer';
import { addFavorite, removeFavorite, addCards, addFortune } from '../../actions';
import save from '../../images/save.png';
import saved from '../../images/saved.png';
import { shallow } from 'enzyme';

describe('FavoriteContainer', () => {

  let wrapper;
  let mockFortune;
  let mockCurrentReading;
  let mockAddReading;
  let mockRemoveReading;

  beforeEach(() => {
    mockFortune = 'Keep your friends close and your enemies closer';
    mockCurrentReading = {
      cards: [{name: 'The Moon', value: 6}], question: 'Is Bithcuits a good kitty?', fortune: 'Keep your friends close and your enemies closer',
      id: 25,
      saved: false
    };
    mockAddReading = jest.fn().mockImplementation();
    mockRemoveReading = jest.fn().mockImplementation();
    wrapper = shallow(<FavoriteContainer
       fortune={mockFortune}
       currentReading={mockCurrentReading}
       addReadingToFavorites={mockAddReading}
       removeReadingFromFavorites={mockRemoveReading}
    />);
  });

  describe('FavoriteContainer container/component', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should start with a default state', () => {
      expect(wrapper.state()).toEqual({icon: save})
    })

    it('should call saveReading if the reading has not been saved yet', () => {
      wrapper.instance().saveReading = jest.fn();
      wrapper.instance().updateSavedStatus();
      expect(wrapper.instance().saveReading).toHaveBeenCalled();
    })

    it('should call removeReading if the reading is already saved', () => {
      mockCurrentReading.saved = true;
      wrapper.instance().removeReading = jest.fn();
      wrapper.instance().updateSavedStatus();
      expect(wrapper.instance().removeReading).toHaveBeenCalled();
    })

    it('should set state, update the current reading\'s saved value, and call addReadingToFavorites when saveReading is invoked', () => {
      const expected = {
        cards: [{name: 'The Moon', value: 6}], question: 'Is Bithcuits a good kitty?', fortune: 'Keep your friends close and your enemies closer',
        id: 25,
        saved: true
      };
      wrapper.instance().saveReading();
      expect(wrapper.state()).toEqual({icon: saved})
      expect(mockAddReading).toHaveBeenCalledWith(expected);
    })

    it('should set state, update the current reading\'s saved value, and call removeReadingFromFavorites when removeReading is invoked ', () => {
      mockCurrentReading.saved = true;
      const expected = {
        cards: [{name: 'The Moon', value: 6}], question: 'Is Bithcuits a good kitty?', fortune: 'Keep your friends close and your enemies closer',
        id: 25,
        saved: false
      };
      wrapper.instance().removeReading();
      expect(wrapper.state()).toEqual({icon: save})
      expect(mockRemoveReading).toHaveBeenCalledWith(expected);
    })

    it('should be able to determine which icon to display', () => {
      expect(wrapper.instance().determineIcon()).toEqual(save);

      mockCurrentReading.saved = true;
      expect(wrapper.instance().determineIcon()).toEqual(saved);
    })

    it('should call updateSavedStatus when the save button is clicked', () => {
      wrapper.instance().updateSavedStatus = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('.save-btn').simulate('click');
      expect(wrapper.instance().updateSavedStatus).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return a current reading object', () => {
      const mockDispatch = jest.fn();
      const mockState = {
        user: {name: 'Bithcuits', id: 9},
        question: 'Is Bithcuits a good kitty?',
        fortune: 'Keep your friends close and your enemies closer',
        currentReading: {
          cards: [{name: 'The Moon', value: 6}],
          question: 'Is Bithcuits a good kitty?',
          fortune: 'Keep your friends close and your enemies closer',
          id: 25,
          saved: false
        }
      }
      const expected = {
        currentReading: {
          cards: [{name: 'The Moon', value: 6}],
          question: 'Is Bithcuits a good kitty?',
          fortune: 'Keep your friends close and your enemies closer',
          id: 25,
          saved: false
        }
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    })
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the addFavorite action when addReadingToFavorites is called', () => {
      const mockFavorite = {cards: [{name: 'The Magician', value: 9}]};
      const mockDispatch = jest.fn();
      const actionToDispatch = addFavorite(mockFavorite);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addReadingToFavorites(mockFavorite);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })

    it('should call dispatch with the removeFavorite action when removeReadingFromFavorites is called', () => {
      const mockFavorite = {cards: [{name: 'The Moon', value: 6}]};
      const mockDispatch = jest.fn();
      const actionToDispatch = removeFavorite(mockFavorite);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.removeReadingFromFavorites(mockFavorite);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  });
});
