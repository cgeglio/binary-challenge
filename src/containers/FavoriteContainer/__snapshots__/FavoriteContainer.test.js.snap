// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteContainer FavoriteContainer container/component should match the snapshot 1`] = `
<section
  className="favorite-container"
>
  <section
    className="cards"
  >
    <Card
      card={
        Object {
          "name": "The Moon",
          "value": 6,
        }
      }
      key="NaN"
    />
  </section>
  <section
    className="reading-details"
  >
    <button
      className="save-btn"
      onClick={[Function]}
    >
      <img
        alt="save reading icon"
        className="save-icon"
        src="save.png"
      />
    </button>
    <div>
      <h2
        className="question"
      >
        Is Bithcuits a good kitty?
      </h2>
      <h2>
        Keep your friends close and your enemies closer
      </h2>
    </div>
  </section>
</section>
`;
